go/test/endtoend/vtgate/partialfailure/main_test.go

```diff
@@ -0,0 +1,561 @@
+			utils.Exec(t, conn, `rollback`)
+			utils.AssertMatches(t, conn, `select id, val1 from test order by id`, `[]`)
+		})
+	}
+}
+
+func TestPartialVindexQueryFailureNoAutoCommitRollback(t *testing.T) {
+	tcases := []struct {
+		mode string
+		qs   []string
+	}{
+		{"oltp", []string{"set workload = oltp"}},
+		{"oltp-reserved", []string{"set workload = oltp", "set sql_mode = ''"}},
+		{"olap", []string{"set workload = olap"}},
+		{"olap-reserved", []string{"set workload = olap", "set sql_mode = ''"}},
+	}
+
+	for _, tc := range tcases {
+		t.Run(tc.mode, func(t *testing.T) {
+			conn, err := mysql.Connect(context.Background(), &vtParams)
+			require.NoError(t, err)
+			defer conn.Close()
+
+			// setup the mode
+			for _, q := range tc.qs {
+				utils.Exec(t, conn, q)
+			}
+
+			// cleanup previous run data from table.
+			utils.Exec(t, conn, `delete from test`)
+
+			utils.Exec(t, conn, `set autocommit = off`)
+			utils.Exec(t, conn, `insert into test(id, val1) values (1,'A'),(2,'B'),(3,'C')`)
+			utils.AssertContainsError(t, conn, `insert into test(id, val1) values (4,'D'),(5,'C')`, `lookup.Create`)
+			utils.AssertMatches(t, conn, `select id, val1 from test order by id`, `[[INT64(1) VARCHAR("A")] [INT64(2) VARCHAR("B")] [INT64(3) VARCHAR("C")]]`)
+			utils.Exec(t, conn, `insert into test(id, val1) values (4,'D'),(5,'E')`)
+			utils.Exec(t, conn, `rollback`)
+			utils.AssertMatches(t, conn, `select id, val1 from test order by id`, `[]`)
+		})
+	}
+}
+
+func TestPartialQueryFailureAutoCommitRollback(t *testing.T) {
+	tcases := []struct {
+		mode string
+		qs   []string
+	}{
+		{"oltp", []string{"set workload = oltp"}},
+		{"oltp-reserved", []string{"set workload = oltp", "set sql_mode = ''"}},
+		{"olap", []string{"set workload = olap"}},
+		{"olap-reserved", []string{"set workload = olap", "set sql_mode = ''"}},
+	}
+
+	for _, tc := range tcases {
+		t.Run(tc.mode, func(t *testing.T) {
+			conn, err := mysql.Connect(context.Background(), &vtParams)
+			require.NoError(t, err)
+			defer conn.Close()
+
+			// setup the mode
+			for _, q := range tc.qs {
+				utils.Exec(t, conn, q)
+			}
+
+			// cleanup previous run data from table.
+			utils.Exec(t, conn, `delete from test`)
+
+			utils.Exec(t, conn, `insert into test(id, val1) values (1,'A'),(2,'B'),(3,'C')`)
+			utils.AssertContainsError(t, conn, `insert into test(id, val1) values (1,'D'),(4,'E')`, `transaction rolled back to reverse changes of partial DML execution`)
+			utils.AssertMatches(t, conn, `select id, val1 from test order by id`, `[[INT64(1) VARCHAR("A")] [INT64(2) VARCHAR("B")] [INT64(3) VARCHAR("C")]]`)
+			utils.Exec(t, conn, `rollback`)
+			utils.AssertMatches(t, conn, `select id, val1 from test order by id`, `[[INT64(1) VARCHAR("A")] [INT64(2) VARCHAR("B")] [INT64(3) VARCHAR("C")]]`)
+		})
+	}
+}
+
+func TestPartialVindexQueryFailureAutoCommitRollback(t *testing.T) {
+	tcases := []struct {
```

Comment:
nit: this pattern is copy pasted over and over again in this file. I think we should extract to a single place